### ISSUE TITLE
syscall: UtimesNano does not use nanosecond system call on BSD, Solaris

### ISSUE THREAD

**Issue by calmh** [golang/go#16480]
ZFS on Solaris retains nanosecond precision file timestamps. These can be seen in the information returned by `os.Stat`. However `os.Chtimes` seems limited to setting microsecond precision. This means, among other things, that a file can't be copied with the timestamp fully retained.

```
jb@zlogin4:~/s/g/s/s/test $ go version
go version go1.7rc3 solaris/amd64
```

```
# Nanoseond precision before running program:
jb@zlogin4:~/s/g/s/s/test $ touch test.file 
jb@zlogin4:~/s/g/s/s/test $ stat test.file
  File: ‘test.file’
  Size: 0               Blocks: 1          IO Block: 131072 regular empty file
Device: 5a00010023h/386547122211d       Inode: 96951       Links: 1
Access: (0644/-rw-r--r--)  Uid: (  502/      jb)   Gid: (    1/   other)
Access: 2016-07-23 20:57:53.896596091 +0200
Modify: 2016-07-23 20:57:53.896596091 +0200
Change: 2016-07-23 20:57:53.896613387 +0200
 Birth: -

# Go program sees correct precision, attempts to set the same:
jb@zlogin4:~/s/g/s/s/test $ go run testchtimes.go test.file
2016-07-23 20:57:53.896596091 +0200 CEST

# Precision was reduced
jb@zlogin4:~/s/g/s/s/test $ stat test.file
  File: ‘test.file’
  Size: 0               Blocks: 1          IO Block: 131072 regular empty file
Device: 5a00010023h/386547122211d       Inode: 96951       Links: 1
Access: (0644/-rw-r--r--)  Uid: (  502/      jb)   Gid: (    1/   other)
Access: 2016-07-23 20:57:53.896596000 +0200
Modify: 2016-07-23 20:57:53.896596000 +0200
Change: 2016-07-23 20:58:02.133653117 +0200
 Birth: -

# The program:
jb@zlogin4:~/s/g/s/s/test $ cat testchtimes.go 
package main

import (
        "fmt"
        "os"
)

func main() {
        info, err := os.Stat(os.Args[1])
        if err != nil {
                fmt.Println(err)
                os.Exit(1)
        }
        t := info.ModTime()
        fmt.Println(t)
        if err := os.Chtimes(os.Args[1], t, t); err != nil {
                fmt.Println(err)
                os.Exit(1)
        }
}
```

**Comment by bradfitz** [comment]
I don't think there's anything Solaris-specific here.

The os package does the same thing on all POSIX systems (file_posix.go):

```go
func Chtimes(name string, atime time.Time, mtime time.Time) error {
        var utimes [2]syscall.Timespec
        utimes[0] = syscall.NsecToTimespec(atime.UnixNano())
        utimes[1] = syscall.NsecToTimespec(mtime.UnixNano())
        if e := syscall.UtimesNano(name, utimes[0:]); e != nil {
                return &PathError{"chtimes", name, e}
        }
        return nil
}
```

... where syscall.UtimesNano always converts to microseconds. Many operating systems have `utimens` etc support but we don't appear to ever use it.

The os.Chtimes documentation is a little misleading:

>    The underlying filesystem may truncate or round the values to a less precise
>    time unit. If there is an error, it will be of type *PathError.

... since it's not really the filesystem truncating things.

This seems fixable.

**Comment by odeke-em** [comment]
Currently only the syscall Linux-es' packages use utimensat in syscall.UtimesNano so that
 nano-second precision is kept for the Linux-es but not other OSes. You can see that this is true on the Linuxes from the playground here https://play.golang.org/p/Jwg_7NW2gx and below from this search
```shell
Emmanuels-MBP-2:syscall emmanuelodeke$ go version
go version devel +5df7f52 Sun Sep 25 23:42:59 2016 +0000 darwin/amd64
Emmanuels-MBP-2:syscall emmanuelodeke$ grep -Rn 'func utimensat' *
zsyscall_linux_386.go:128:func utimensat(dirfd int, path string, times *[2]Timespec) (err error) {
zsyscall_linux_amd64.go:128:func utimensat(dirfd int, path string, times *[2]Timespec) (err error) {
zsyscall_linux_arm.go:128:func utimensat(dirfd int, path string, times *[2]Timespec) (err error) {
zsyscall_linux_arm64.go:128:func utimensat(dirfd int, path string, times *[2]Timespec) (err error) {
zsyscall_linux_mips64.go:126:func utimensat(dirfd int, path string, times *[2]Timespec) (err error) {
zsyscall_linux_mips64le.go:126:func utimensat(dirfd int, path string, times *[2]Timespec) (err error) {
zsyscall_linux_ppc64.go:128:func utimensat(dirfd int, path string, times *[2]Timespec) (err error) {
zsyscall_linux_ppc64le.go:128:func utimensat(dirfd int, path string, times *[2]Timespec) (err error) {
zsyscall_linux_s390x.go:126:func utimensat(dirfd int, path string, times *[2]Timespec) (err error) {
```

Unfortunately doesn't seem like darwin doesn't support utimensat and there is a comment in the syscall code about it
<img width="865" alt="screen shot 2016-09-27 at 1 06 43 am" src="https://cloud.githubusercontent.com/assets/4898263/18865152/b305a9a6-844e-11e6-907f-3437c0e40ec3.png">



However, it seems like we can do this for Solaris and perhaps the other BSDs.
@mdempsky could you help me confirm whether this is still true and if utimensat can be  supported on the other BSDs?

**Comment by rsc** [comment]
I think the package docs for os.Chtimes are correct. For example you can use a nanosecond-specific system call but if you're writing to FAT32, it's still going to get truncated to 2-second granularity.

The expectation is that the implementation of syscall.UtimesNano uses the most precise system call available. That's true on Linux: it tries utimensat and falls back to utimes when utimensat is not available. It appears to be true on Windows too: it uses SetFileTime which works at 100ns granularity, but that's true of everything on Windows.

On BSD systems, there is a TODO in syscall_bsd.go to use a higher precision call.

On Solaris, there is no TODO but still the implementation uses plain syscall.Utimes. If Solaris has a more precise call, the Solaris implementation of syscall.UtimesNano should use it.

**Comment by mdempsky** [comment]
@odeke-em utimensat has been supported since OpenBSD 5.0 (Nov 2011), NetBSD 6.0 (Oct 2012), and FreeBSD 11.0 (Oct 2016).  It looks like it's still unsupported on Darwin though.

**Comment by odeke-em** [comment]
Thanks for the clarification @mdempsky!

@bradfitz I don't know who to ask but might you access to some Solaris and BSD machines?

**Comment by odeke-em** [comment]
Actually ama let someone who is an expert on syscalls and BSD handle this,
instead of me holding up the freeze cut-off that is in less than 30 hours.

**Comment by binarycrusader** [comment]
All actively supported versions of Solaris from Solaris 10 onwards support utimensat since roughly 2010.

**Comment by binarycrusader** [comment]
You can see that I added support for this in the sys/unix package as part of the changes for #8609:

https://github.com/golang/sys/commit/68a71b6be5774eece23e35475fc28514a84297a3

**Comment by binarycrusader** [comment]
FreeBSD actually first added support for utimensat in FreeBSD 10 as part of this review in January 2015:

https://reviews.freebsd.org/D1426

http://fxr.watson.org/fxr/source/sys/stat.h?v=FREEBSD10#L331

https://www.freebsd.org/cgi/man.cgi?query=utimensat&apropos=0&sektion=0&manpath=FreeBSD+10.3-RELEASE+and+Ports&arch=default&format=html

**Comment by odeke-em** [comment]
Hello there @binarycrusader, the Go1.9 tree is open. Any interest/bandwidth for getting this in? Thanks.

**Comment by tklauser** [comment]
The same problem also exists in x/sys/unix. I'll have a look at it and send CLs for both.

**Comment by tklauser** [comment]
Submitted https://golang.org/cl/55130 attempting to fix this issue in the `syscall` package. So far, I was only able to do some rudimentary testing on FreeBSD so far and I'd appreciate if anyone running Solaris with a nanosecond-precision file system could give it a try (/cc @calmh @binarycrusader). The change is also available in tklauser/go@a17c114.
